java/java-ipfs.md

```diff
@@ -0,0 +1,485 @@
+
+Let's say for example we want to host an article on IPFS, the article (version 1) will have a unique hash, but if we decide to update this article and host it on IPFS, the hash will be competely different and we will have to reshare the new hash. IPNS can be used to prevent this issue, it is possible to link a name to a hash and update the hash a much as we want so we only have to reassign the hash of the article to a name, and share the name, if we update the article, we only have to update the name resolution to point to the latest version.
+
+*Note: IPNS is still very much work in progress and is very slow in practice, it will take approxmitly 1-2 min to publish a name.*
+
+### Keys
+
+IPNS is a based on a distributed Public Key Infrastructure (PKI). So to get started, it is required to have a keypair available on our IPFS node.
+
+A Keypair can be used to store one key/value pair where the key represents the "name" and "value" the hash to resolve.
+
+#### Generate a key
+
+First of all, we need to generate a keypair using the method `ipfs.key.gen(name, type, size): KeyInfo`.
+
+```java
+String keyName ="myarticle";
+Optional<String> keyType = Optional.of("rsa");
+Optional<String> keySize = Optional.of("2048");
+
+KeyInfo key = ipfs.key.gen(keyName, keyType, keySize);
+System.out.println("key name: " + key.name);
+System.out.println("key.hash: " + key.id);
+```
+
+![](https://i.imgur.com/KQUKhk2.png)
+
+The following function returns a `KeyInfo` object composed of the name and the id (multihash) of the key representing the name which can be used to resolve a hash.
+
+#### Delete a key
+
+Alternatively, it is possible to remove a key using `ipfs.key.rm(keyName): void`.
+
+```java
+ipfs.key.rm(keyName);
+```            
+
+#### List all keys
+
+The method `ipfs.key.list()` allows to list all keys available on our node.
+
+```java
+List<KeyInfo> keys = ipfs.key.list();
+keys.forEach(key ->
+    System.out.println("keyInfo: name=" + key.name + ", hash=" + key.id));			
+```
+
+![](https://i.imgur.com/3TefgpL.png)
+
+The `self` key represents the default key generated when we launch IPFS for the first time.
+
+
+### Publish
+
+Once we have a exclusive keypair available, we can use it to publish a hash against it using the method `ipfs.name.publish(hash, keyName)`:
+
+```java
+String hash = "QmWfVY9y3xjsixTgbd9AorQxH7VtMpzfx2HaWtsoUYecaX" // Hash of "hello
+Map response = ipfs.name.publish(hash, Optional.of(keyName));
+System.out.println("publish(hash="+hash+", key="+keyName+"): " + response);
+```      
+
+![](https://i.imgur.com/vRIeeWF.png)
+
+*Note: this operation is particularly slow and can take up to two minutes to execute*
+
+
+### Resolve
+
+Just like a DNS, reading an Object from an IPNS name is a two steps process:
+1. Resolve the hash against the name
+2. Read the content from the hash
+
+```java
+String resolveResponse = ipfs.name.resolve(key.id);
```

Comment:
@gjeanmart Should this code example be in a loop or something? i.e. what is `key` in this case?

Reply:
`key` is the key generated at the step "Generate a key". we can indeed get the key by name via a loop
```java
try {
  KeyInfo keyInfo = localIPFS.key.list().stream()
    .filter(key -> key.name.equals(keyName))
    .findAny()
    .orElseThrow(() -> new RuntimeException("Key " + keyName + " not found"));

  String resolveResponse = ipfs.name.resolve(key.id);
  System.out.println("resolve(key="+key.id+"): " + resolveResponse);

  byte[] content = ipfs.cat(Multihash.fromBase58(resolveResponse.substring(6)));
  System.out.println("Content: " + new String(content));
} catch (IOException ex) {
 throw new RuntimeException("Error whilst communicating with the IPFS node", ex);
}